File: Wed 2/regularization.py
# Regularization

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].to_numpy()
y = dataset.iloc[:, -1].to_numpy()

# Encoding categorical data
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [3])],
                       remainder='passthrough')
X = ct.fit_transform(X)

# Avoiding the Dummy Variable Trap
X = X[:, 1:]

# Feature Scaling: scale your data
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X[:, 2:] = sc.fit_transform(X[:, 2:])
y = sc.fit_transform(y.reshape(len(y), 1)).flatten()

# Fitting Multiple Linear Regression with Regularization: use this after scaling 
# alpha is the regularization parameter
# we want 
from sklearn.linear_model import ElasticNet
regressor1 = ElasticNet(l1_ratio = 0.5, alpha = 0)
regressor1.fit(X, y)
print(regressor1.coef_)

regressor2 = ElasticNet(l1_ratio = 0.5, alpha = 1)
regressor2.fit(X, y)
print(regressor2.coef_)

File: Evaluating Regression Model Performance/regularization.R
# Regularization

# Importing the dataset
dataset = read.csv('50_Startups.csv')

# Feature Scaling
dataset[, c(1:3, 5)] = scale(dataset[, c(1:3, 5)])

# Preparing independent variable matrix
X = model.matrix(Profit ~ ., dataset)[, 2:6]

# Preparing dependent variable vector
y = dataset$Profit

# Fitting Multiple Linear Regression with Regularization
library(glmnet)
regressor1 = glmnet(X, y, alpha = 0.5, lambda = 0)
print(coef(regressor1))
regressor2 = glmnet(X, y, alpha = 0.5, lambda = 1)
print(coef(regressor2))